### PR TITLE
sync raft log

### PR DESCRIPTION
**Pull request by javaforfun**
For high reliability, we need sync raft log when append.  This can prevent data loss when power failure.

### CODE REVIEW

File: etc/config-template.toml [hunk 1]
```diff
@@ -41,6 +41,9 @@ address = ""
 job = "tikv"
 
 [raftstore]
+# Set to true if high reliability needed, eg. prevent data loss when power failure.
+raft-log-synchronize = false
```
Comment (siddontang) [inline etc/config-template.toml]: sync-log = true

File: src/raft/raw_node.rs [hunk 2]
```diff
@@ -111,6 +111,10 @@ pub struct Ready {
     // If it contains a MsgSnap message, the application MUST report back to raft
     // when the snapshot has been received or has failed by calling ReportSnapshot.
     pub messages: Vec<Message>,
+
+    // MustSync indicates whether the HardState and Entries must be synchronously
+    // written to disk or if an asynchronous write is permissible.
+    pub must_sync: bool,
```
Comment (siddontang) [inline src/raft/raw_node.rs]: we don't need this field, we will force syncing raft log all time.

File: etc/config-template.toml (hunk 1 above)
Comment (zhangjinpeng1987) [inline etc/config-template.toml]: `# sync-raft-log = false`

**Comment by zhangjinpeng1987** [comment]
Please add bench result for open `sync-raft-log`.

File: etc/config-template.toml [hunk 3]
```diff
@@ -41,6 +41,9 @@ address = ""
 job = "tikv"
 
 [raftstore]
+# Set to true if high reliability needed, eg. prevent data loss when power failure.
+# sync-log = false
```
Comment (ngaut) [inline etc/config-template.toml]: We should set the default value to true in case of machine failure.

File: src/bin/tikv-server.rs [hunk 4]
```diff
@@ -607,6 +607,7 @@ fn build_cfg(matches: &ArgMatches,
     cfg_usize(&mut cfg.send_buffer_size, config, "server.send-buffer-size");
     cfg_usize(&mut cfg.recv_buffer_size, config, "server.recv-buffer-size");
 
+    cfg.raft_store.sync_log = get_toml_boolean(config, "raftstore.sync-log", Some(false));
```
Comment (ngaut) [inline src/bin/tikv-server.rs]: Ditto

File: src/raftstore/store/store.rs [hunk 5]
```diff
@@ -898,7 +899,9 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         self.raft_metrics.ready.has_ready_region += append_res.len() as u64;
 
         if !wb.is_empty() {
-            self.engine.write(wb).unwrap_or_else(|e| {
+            let mut write_opts: WriteOptions = WriteOptions::new();
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/store.rs]: No need specified the type of `WriteOptions`.

**Comment by siddontang** [comment]
PTAL @ngaut

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by ngaut** [comment]
PTAL @BusyJay

File: etc/config-template.toml [hunk 6]
```diff
@@ -49,6 +49,9 @@ address = ""
 job = "tikv"
 
 [raftstore]
+# true(default value) for high reliability, this can prevent data loss when power failure.
```
Comment (BusyJay) [inline etc/config-template.toml]: Leave a space before `(`.

File: src/raftstore/store/config.rs [hunk 7]
```diff
@@ -138,6 +141,7 @@ pub struct Config {
 impl Default for Config {
     fn default() -> Config {
         Config {
+            sync_log: false,
```
Comment (BusyJay) [inline src/raftstore/store/config.rs]: I don't think it's a good idea to set the default value to false. And it's conflict with the [description](https://github.com/pingcap/tikv/pull/1840/files#diff-d4c79fcc9faffa48ae170c90d8c0683aR52).

File: src/raft/raw_node.rs (hunk 2 above)
Comment (anan-cn) [reply src/raft/raw_node.rs]: I wonder why we need this 'sync-log = true'?

**Pull request merged**